src/sfCommand.ts

```diff
@@ -33,14 +33,20 @@ export abstract class SfCommand<T> extends Command {
   public static envVariablesSection?: HelpSection;
   public static errorCodes?: HelpSection;
   public static tableFlags = CliUx.ux.table.flags;
+  public static requiresProject: boolean;
```

Comment:
What do you think of keep this more specific (e.g. `requiresSfdxProject` or even `requiresLegacyProject`) until we know what a Salesforce project will be in a "unified world".

Same for the references to `project` below.

Reply:
@RodEsp I'm not opposed to it but I think it's going to be too hard to predict what the future of salesforce projects will be. So I'm leaning towards just doing this for now and we can handle any changes later

Reply:
@mdonnalley that's why I'm suggesting this. If we just name this "project" and Salesforce projects end up significantly changing what "project" means in our code this will be an issue. But if we name this "sfdxProject" or something more specific it won't be an issue no matter what a Salesforce project turns out to be.

Reply:
@RodEsp My assumption is that the future of "projects" will be a superset of what an sfdx project currently is - not something entirely different. Based on that assumption, I think it will be easier to update the underlying project class than it would be to go update dozens of commands that now would need to use `requiresProject` instead of `requiresSfdxProject`

Reply:
@mdonnalley I guess I see that similarly but kind of in the opposite way. I agree that the future of "projects" will be a superset of what an sfdx project is right now but that means that `requiresProject` will at some point encompass more than what an sfdx project is, and therefore any command that _only_ requires an sfdx project, or that _specifically_ requires an sfdx project, will now need to go change `requiresProject` to some newly created property, like `requiresLegacyProject`.

I think commands that can do things with the future "Salesforce projects" will probably need to be changed anyways to accommodate them and changing one property (`requiresSfdxProject` -> `requiresProject`) will be the least of it.

But I'm fine leaving this as is if you feel strongly about it.

Reply:
@mdonnalley @RodEsp would any of the above concerns be eased by changing `requiresProject` type to an enum?

enum ProjectType {
  'sfdx',
  'sf'
}

requiresProject: ProjectType;

OR

requiresProject: ProjectType[]; // to prepare for complex projects of the future

Reply:
I like `requiresProject: ProjectType[]`. Although that would require further logic to determine what an `sfdx` project looks like vs an `sf` (or w/e) project looks like.

I haven't run this code but just using it as an example:
```js
public async _run<R>(): Promise<R | undefined> {
  if (this.statics.requiresProject) {
    this.project = await this.assignProject(this.statics.requiresProject);
  }
  return super._run<R>();
}

protected async assignProject(type: ProjectType): Promise<SfdxProject> {
  try {
    if (type === ProjectType.SFDX) {
      return await SfdxProject.resolve();
    }
  } catch (err) {
    if (err instanceof Error && err.name === 'InvalidProjectWorkspaceError') {
      throw messages.createError(`This command is required to be run from a directory containing a ${type} project.`);
    }
    throw err;
  }
}
```